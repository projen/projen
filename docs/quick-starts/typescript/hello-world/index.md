---
sidebar_position: 1
---

# Hello World with TypeScript

To create a new TypeScript project, use `npx projen new typescript`:

```shell
npx projen new typescript
```

This will create a `.projenrc.ts` file which defines your project.

`npx projen new` will synthesize a standard project directory structure with some sample
code.

```shell
├── src
│   ├── index.ts
├── test
│   ├── hello.test.ts
├── node_modules
├── package.json
└── tsconfig.json
```

The default options will set up your `tsconfig` root directory as `src` and your
out directory as `lib`. By default the `test` directory is not compiled to
JavaScript and instead compiled when `npx projen test` is executed.

The `npx projen new` command will also generate a `.projenrc.js` file which includes
the definition of your project with any options you specified in the command
line:

```js
const { typescript } = require("projen");
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: "main",
  name: "my-project",
});
project.synth();
```

To modify your project definitions, edit `.projenrc.ts` and run `projen` again
to re-synthesize your project. The following sections describe the various
features of your TypeScript project.

## Dependencies

You can specify dependencies in your project via the `deps`, `devDeps`, and
`peerDeps` options similar to what you might expect in a `package.json` file.

The recommendation is to only specify the module name here (e.g. `express`).
This will behave similar to yarn add or npm install in the sense that it will
add the module as a dependency to your package.json file with the latest version
(`^`). You can specify semver requirements in the same syntax passed to npm i or
yarn add (e.g. `express@^2`) and this will be what your `package.json` will
eventually include.

## Migrating your TypeScript Project to Projen

Projen provides some opinionated defaults that may be different than the defaults
you rely on in your existing TypeScript Project. This section attempts to document
some of the common themes to look out for when migrating an existing TypeScript
project to Projen.

## Migrating .projenrc.js to .projenrc.ts

If you'd like to upgrade an existing project that was creating using a JS-based
projen RC file (.projenrc.js) to a TS-based projen RC file (.projenrc.ts):

1. Add `projenrcTs: true` to your project.
2. Run `npx projen`.
3. Rename `.projenrc.js` to `.projenrc.ts`.
4. Update `require`s to `import`s.
5. Run `npx projen`.

### Default Directory Structure

The most important is the default project structure. Projen expects all source code
to be in the `src` directory and will write compiled `.js` and `.d.ts` files to the
`lib` directory. Test files are expected in the `test` directory and not compiled
to javascript. Instead, Projen configures `Jest` to compile `.ts` files when testing.

As this is only one of many structures for valid TypeScript projects, your migrated
project may not work out-of-the-box with Projen. We recommend that you take the
time to fully migrate to the expected project structure so that you can take full
advantage of Projen management on your repository.

However, you can also configure your `.projenrc.js` file to reflect your current
project structure. A `TypeScriptProject` exposes all options in a `tsconfig` file
so you can manually set the structure you want. Here is an example for specifying a
structure where the TypeScript compiler finds all TypeScript files recursively in your
repository:

```js
const { typescript } = require('projen');
const project = new typescript.TypeScriptProject({
  ...
  tsconfig: {
    compilerOptions: {
      rootDir: '.',
      outDir: '.',
    },
    include: ['**/*.ts'], // all typescript files recursively
  },
});
```

### Dependency Versions

As described in the section above, Projen recommends that you list your dependencies
only by module name and have Projen install the latest version of the package. A
consequence of this recommendation is that when migrating, you may unknowingly update
your dependencies to incompatible versions. You can always provide specific semver
requirements (e.g. `express@2.1.0`) if necessary.
