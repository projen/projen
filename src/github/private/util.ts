export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
