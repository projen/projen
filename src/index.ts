export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
