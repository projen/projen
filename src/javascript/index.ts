export * from "./bundler";
export * from "./eslint";
export * from "./jest";
export * from "./license-checker";
export * from "./node-package";
export * from "./node-project";
export * from "./npm-config";
export * from "./prettier";
export * from "./projenrc";
export * from "./typescript-config";
export * from "./upgrade-dependencies";
export * from "./yarnrc";
