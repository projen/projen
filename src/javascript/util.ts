import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
