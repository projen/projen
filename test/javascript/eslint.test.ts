import { TaskRuntime } from "../../src";
import { Eslint, NodeProject } from "../../src/javascript";
import { TypeScriptProject } from "../../src/typescript";
import { execProjenCLI, synthSnapshot } from "../util";

test.each([
  ["prettier is off", false],
  ["prettier is on", true],
])("eslint task passes with default config: %s", (_, prettier) => {
  const project = new TypeScriptProject({
    name: "test",
    defaultReleaseBranch: "master",
    eslint: true,
    prettier,
  });
  project.synth();

  // THEN
  execProjenCLI(project.outdir, ["eslint"]);
});

test("can acceess file", () => {
  // GIVEN
  const project = new NodeProject({
    name: "test",
    defaultReleaseBranch: "master",
  });

  const eslint = new Eslint(project, {
    dirs: ["mysrc"],
    lintProjenRc: false,
  });

  // WHEN
  eslint.file.addOverride("env.foo", "bar");

  // THEN
  const output = synthSnapshot(project)[".eslintrc.json"];
  expect(output.env).toHaveProperty("foo", "bar");
});

describe("prettier", () => {
  test("snapshot", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    new Eslint(project, {
      dirs: ["mysrc"],
      prettier: true,
      lintProjenRc: false,
    });

    // THEN
    expect(synthSnapshot(project)[".eslintrc.json"]).toMatchSnapshot();
  });
});

describe("alias", () => {
  test("custom config", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["mysrc"],
      aliasMap: {
        "@src": "./src",
        "@foo": "./src/foo",
      },
      aliasExtensions: [".ts", ".js"],
      lintProjenRc: false,
    });

    // THEN
    expect(eslint.config.settings["import/resolver"].alias).toHaveProperty(
      "map",
      [
        ["@src", "./src"],
        ["@foo", "./src/foo"],
      ]
    );
    expect(eslint.config.settings["import/resolver"].alias).toHaveProperty(
      "extensions",
      [".ts", ".js"]
    );
  });
});

describe("eslint settings", () => {
  test("devdirs", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    new Eslint(project, {
      devdirs: ["foo", "bar"],
      dirs: ["mysrc"],
      lintProjenRc: false,
    });

    // THEN
    expect(synthSnapshot(project)[".eslintrc.json"]).toMatchSnapshot();
  });

  test("tsAlwaysTryTypes", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["mysrc"],
      tsAlwaysTryTypes: true,
      lintProjenRc: false,
    });

    // THEN
    expect(eslint.config.settings["import/resolver"].typescript).toHaveProperty(
      "alwaysTryTypes",
      true
    );
  });

  test("if the prettier is configured, eslint is configured accordingly", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
    });

    eslint.addExtends("plugin:some-plugin/recommended");

    // THEN
    const output = synthSnapshot(project);

    // Prettier should be last in the extends array
    const extendsArray = output[".eslintrc.json"].extends;
    expect(extendsArray).toEqual([
      "plugin:import/typescript",
      "plugin:some-plugin/recommended",
      "plugin:prettier/recommended",
    ]);
  });

  test("not setting sortExtends should correctly produce the default order", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
    });

    // Add the prettier plugins in the incorrect order
    eslint.addExtends("prettier");
    eslint.addExtends("plugin:prettier/recommended");

    // Add some other plugins
    eslint.addExtends("plugin:some-plugin/recommended");
    eslint.addExtends("plugin:a-second-plugin/recommended");

    // THEN
    const output = synthSnapshot(project);
    const extendsArray = output[".eslintrc.json"].extends;

    expect(extendsArray).toEqual([
      "plugin:import/typescript", // always added

      // Should stay in order they were added into
      "plugin:some-plugin/recommended",
      "plugin:a-second-plugin/recommended",

      // ordered according to best practices
      "plugin:prettier/recommended",
      "prettier",
    ]);
  });

  test("setting sortExtends to a meaningless comparer should leave the extends array order alone", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
      sortExtends: {
        compare: () => 0,
      },
    });

    eslint.addExtends("plugin:some-plugin/recommended");

    // THEN
    const output = synthSnapshot(project);

    const extendsArray = output[".eslintrc.json"].extends;

    expect(extendsArray).toEqual([
      "plugin:import/typescript",
      "plugin:prettier/recommended",
      "plugin:some-plugin/recommended",
    ]);
  });

  test("setting sortExtends to a comparer should use that to sort the extends array", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
      sortExtends: {
        // Backwards alphanumeric
        compare: (a, b) => b.localeCompare(a),
      },
    });

    eslint.addExtends("plugin:some-plugin/recommended");

    // THEN
    const output = synthSnapshot(project);

    const extendsArray = output[".eslintrc.json"].extends;

    expect(extendsArray).toEqual([
      "plugin:some-plugin/recommended",
      "plugin:prettier/recommended",
      "plugin:import/typescript",
    ]);
  });

  test("can output yml instead of json", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "main",
      prettier: true,
    });

    // WHEN
    new Eslint(project, {
      dirs: ["src"],
      yaml: true,
      lintProjenRc: false,
    });

    // THEN
    const output = synthSnapshot(project);
    expect(output[".eslintrc.yml"]).toBeDefined();
    expect(output[".eslintrc.json"]).toBeUndefined();
  });

  test("can override the parser", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
    });
    eslint.addOverride({
      files: ["*.json", "*.json5", "*.jsonc"],
      parser: "jsonc-eslint-parser",
    });
    const output = synthSnapshot(project);

    // THEN
    expect(output[".eslintrc.json"].overrides).toContainEqual({
      files: ["*.json", "*.json5", "*.jsonc"],
      parser: "jsonc-eslint-parser",
    });
  });

  test("creates a eslint task", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
    });

    // THEN
    const manifest = synthSnapshot(project)[TaskRuntime.MANIFEST_FILE];
    expect(eslint.eslintTask._renderSpec()).toMatchObject(
      manifest.tasks.eslint
    );
  });

  test("excludes --fix flag when fix is disabled", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      commandOptions: { fix: false },
    });

    // THEN
    const taskStep = eslint.eslintTask.steps[0];
    expect(taskStep.exec).not.toContain("--fix");
    expect(taskStep?.args ?? []).not.toContain(
      expect.stringContaining("--fix")
    );
  });

  test("omit --ext when no extensions are specified", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
      fileExtensions: [],
    });

    // THEN
    const taskStep = eslint.eslintTask.steps[0];
    expect(taskStep.exec).not.toContain("--ext");
    expect(taskStep?.args ?? []).not.toContain(
      expect.stringContaining("--ext")
    );
  });

  test("add --ext when extensions are specified", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
    });

    // THEN
    const taskStep = eslint.eslintTask.steps[0];
    expect(taskStep.exec).toContain("--ext");
  });

  test("supports specifying extra task args", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      commandOptions: { extraArgs: ["--cache"] },
    });

    // THEN
    const taskStep = eslint.eslintTask.steps[0];
    expect(taskStep.exec).toContain("--cache");
  });

  test("allow modification of the eslint task", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
    });

    const taskStep = eslint.eslintTask.steps[0];
    const newTestArg = "--foo";
    eslint.eslintTask.reset(taskStep.exec, { args: [newTestArg] });

    eslint.addLintPattern("bar");

    // THEN
    expect(eslint.eslintTask.steps[0].args).toContain(newTestArg);
  });
});
