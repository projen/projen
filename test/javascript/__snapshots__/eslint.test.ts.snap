// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eslint settings devdirs 1`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/typescript",
  ],
  "ignorePatterns": [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "coverage",
    "!.projenrc.js",
  ],
  "overrides": [
    {
      "files": [
        ".projenrc.js",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
    "import",
    "@stylistic",
  ],
  "root": true,
  "rules": {
    "@stylistic/array-bracket-newline": [
      "error",
      "consistent",
    ],
    "@stylistic/array-bracket-spacing": [
      "error",
      "never",
    ],
    "@stylistic/brace-style": [
      "error",
      "1tbs",
      {
        "allowSingleLine": true,
      },
    ],
    "@stylistic/comma-dangle": [
      "error",
      "always-multiline",
    ],
    "@stylistic/comma-spacing": [
      "error",
      {
        "after": true,
        "before": false,
      },
    ],
    "@stylistic/indent": [
      "error",
      2,
    ],
    "@stylistic/key-spacing": [
      "error",
    ],
    "@stylistic/keyword-spacing": [
      "error",
    ],
    "@stylistic/max-len": [
      "error",
      {
        "code": 150,
        "ignoreComments": true,
        "ignoreRegExpLiterals": true,
        "ignoreStrings": true,
        "ignoreTemplateLiterals": true,
        "ignoreUrls": true,
      },
    ],
    "@stylistic/member-delimiter-style": [
      "error",
    ],
    "@stylistic/no-multi-spaces": [
      "error",
      {
        "ignoreEOLComments": false,
      },
    ],
    "@stylistic/no-multiple-empty-lines": [
      "error",
    ],
    "@stylistic/no-trailing-spaces": [
      "error",
    ],
    "@stylistic/object-curly-newline": [
      "error",
      {
        "consistent": true,
        "multiline": true,
      },
    ],
    "@stylistic/object-curly-spacing": [
      "error",
      "always",
    ],
    "@stylistic/object-property-newline": [
      "error",
      {
        "allowAllPropertiesOnSameLine": true,
      },
    ],
    "@stylistic/quote-props": [
      "error",
      "consistent-as-needed",
    ],
    "@stylistic/quotes": [
      "error",
      "single",
      {
        "avoidEscape": true,
      },
    ],
    "@stylistic/semi": [
      "error",
      "always",
    ],
    "@stylistic/space-before-blocks": [
      "error",
    ],
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": "error",
    "@typescript-eslint/no-require-imports": "error",
    "@typescript-eslint/no-shadow": "error",
    "@typescript-eslint/return-await": "error",
    "curly": [
      "error",
      "multi-line",
      "consistent",
    ],
    "dot-notation": [
      "error",
    ],
    "import/no-duplicates": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/foo/**",
          "**/bar/**",
        ],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "no-bitwise": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "node": {},
      "typescript": {
        "alwaysTryTypes": true,
        "project": "./tsconfig.json",
      },
    },
  },
}
`;

exports[`prettier snapshot 1`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/typescript",
    "plugin:prettier/recommended",
  ],
  "ignorePatterns": [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "coverage",
    "!.projenrc.js",
  ],
  "overrides": [
    {
      "files": [
        ".projenrc.js",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
    "import",
  ],
  "root": true,
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": "error",
    "@typescript-eslint/no-require-imports": "error",
    "@typescript-eslint/no-shadow": "error",
    "@typescript-eslint/return-await": "error",
    "curly": [
      "error",
      "multi-line",
      "consistent",
    ],
    "dot-notation": [
      "error",
    ],
    "import/no-duplicates": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "no-bitwise": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "node": {},
      "typescript": {
        "alwaysTryTypes": true,
        "project": "./tsconfig.json",
      },
    },
  },
}
`;
